### PR TITLE
Log record mutations are visible in next registered processors

### PR DESCRIPTION
**Pull request by pellared**
### Current description

Closes https://github.com/open-telemetry/opentelemetry-specification/issues/4065 

After few SIG meetings, many discussions, the Go SIG decided to follow the intention of the specification:
- https://github.com/open-telemetry/opentelemetry-go/pull/5636

I propose to clarify the specification to avoid confusion of future readers.

<details>

<summary>Previous description</summary>

Per https://github.com/open-telemetry/opentelemetry-specification/issues/4065#issuecomment-2147846667

Allow chaining processors and having the log record modifications local.
This is how currently [C++](https://github.com/open-telemetry/opentelemetry-specification/issues/4010#issuecomment-2096577303) (stable), Go (beta) Logs SDK are designed.
For reference, the following SDKs implements shared mutations for log record processors: Java, .NET, Python, JavaScript, PHP.

For Go, it allows high performance by minimizing the amount of heap allocations while keeping the design clean and Go idiomatic (similar to structured logging in Go standard library).
https://github.com/open-telemetry/opentelemetry-go/pull/5470 is a PoC of making log record mutation visible in next registered processors but suffers from very awkward design (still it is not compliant with the way the specification is written right now).
https://github.com/open-telemetry/opentelemetry-go/pull/5478 is another PoC of making log record mutation visible in next registered processors, but it suffers more heap allocations and adds synchronization (in my opinion, unacceptable performance overhead).

EDIT:
A shared log record means that consecutive processors receive the same instance of the log record.
Not shared means that each log record record receives a copy of log record.

Some implementations already use a shared record (Java, .NET, Python, JavaScript, Python, Rust). However, in Go (C++ probably as well) we prefer to use copies (shallow) to reduce the amount of heap allocations (makes it possible to have a zero allocation log processing). E.g. it can drastically improve the performance as e.g. trace and debug logs (if info log level is set) would reduce the memory pressure and make the garbage collection a lot more efficient (for 95% of the cases it would not cause any heap allocation).

Even if the other implementations would benefit from using copies as well may not to want to change it as it would be a breaking change.

In short, stack allocated log records can be more performant. Having log records shared causes the Go escape analysis to make a heap allocation more often.

</details>

### CODE REVIEW

**Comment by pellared** [comment]
I will have limited availability for next 3 weeks (almost none for next 2 weeks).

I am totally fine (and even happy) if someone else (maybe @jmacd) would own the PR (e.g. address comments) to keep the momentum.

File: specification/logs/sdk.md [hunk 1]
```diff
@@ -317,9 +317,22 @@ therefore it SHOULD NOT block or throw exceptions.
 
 **Returns:** `Void`
 
-A `LogRecordProcessor` may freely modify `logRecord` for the duration of
-the `OnEmit` call. If `logRecord` is needed after `OnEmit` returns (i.e. for
-asynchronous processing) only reads are permitted.
+It is implementation specific whether and how the `logRecord` passed by the SDK
+is shared between the processors. Log record mutations do not have to be visible
+in next registered processors.
```
Comment (jmacd) [inline specification/logs/sdk.md:322]: Is there an existing SDK that implements shared mutations for log record processors? It seems unnecessarily permissive to support both mutable and immutable modes in the specificaiton. The PR description states that some SDKs do not support mutation. Which do?

Comment (pellared) [reply specification/logs/sdk.md:322]: Java, .NET, Python, JavaScript, PHP

Comment (pellared) [reply specification/logs/sdk.md:322]: I updated the description.

**Comment by pellared** [comment]
@open-telemetry/go-approvers @open-telemetry/cpp-approvers @open-telemetry/rust-approvers PTAL

**Review by XSAM — COMMENTED** [review]
I am confused by the shared `logRecord`. How do we define a shared `logRecord`? What is the use case for the shared `logRecord` and not shared `logRecord`?

Should we provide more specific examples for these two cases to make it clear?

**Comment by pellared** [comment]
@XSAM

A shared log record means that consecutive processors receive the same instance of the log record.

Not shared means that each log record record receives a copy of log record.

Some implementations already use a shared record (Java, .NET, Python, JavaScript, Python). However in Go (C++ and Rust probably as well) we prefer to use copies to reduce the amount of heap allocations. Even if the other implementations would benefit from using copies as well may not to want to change it as it would be a breaking change.

**Review by XSAM — APPROVED** [review]

File: specification/logs/sdk.md [hunk 2]
```diff
@@ -317,9 +317,22 @@ therefore it SHOULD NOT block or throw exceptions.
...
+
+If the `logRecord` is shared, then a `LogRecordProcessor` may freely modify
```
Comment (XSAM) [inline specification/logs/sdk.md:324]:
```suggestion
If the `logRecord` is shared[^shared], then a `LogRecordProcessor` may freely modify
```

File: specification/logs/sdk.md [hunk 3]
```diff
@@ -317,9 +317,22 @@ therefore it SHOULD NOT block or throw exceptions.
...
+`logRecord` only for the duration of the `OnEmit` call.
+The modifications applied on shared data are visible in next
+registered processors.
+If `logRecord` is needed after `OnEmit` returns (i.e. for asynchronous
+processing) only reads are permitted.
+
+If the `logRecord` is not shared, then modifications applied on the `logRecord`
```
Comment (XSAM) [inline specification/logs/sdk.md:331]:
```suggestion
If the `logRecord` is not shared[^not-shared], then modifications applied on the `logRecord`
```

File: specification/logs/sdk.md [hunk 4]
```diff
@@ -317,9 +317,22 @@ therefore it SHOULD NOT block or throw exceptions.
...
+are local.
+If the modification is needed in another processor,
+then the processor making the modification needs to call the other processor's
+`OnEmit` and pass the locally modified `logRecord`.
```
Comment (XSAM) [inline specification/logs/sdk.md:336]:
```suggestion

[^shared]: A shared `logRecord` means that consecutive processors receive the same instance of the log record. (i.e. shared by a pointer)
[^not-shared]: A not shared `logRecord` means that consecutive processors receive a copy of the log record. (i.e. shallow copies in Go, which are stack allocations)
```

**Review by XSAM — COMMENTED** [review]
We can leave the definition of shared `logRecord` in the doc, so future viewers don't need to check the PR to know the definition.

File: specification/logs/sdk.md [hunk 5]
```diff
@@ -317,9 +317,29 @@ therefore it SHOULD NOT block or throw exceptions.
...
+the same instance of the log record (i.e. shared by a pointer).
+
+If the `logRecord` is not shared[^not-shared], then modifications applied
+on the `logRecord` are local.
+If the modification is needed in another processor,
+then the processor making the modification needs to call the other processor's
+`OnEmit` and pass the locally modified `logRecord`.
+
+[^not-shared]: A not shared `logRecord` means that consecutive processors
+receive a copy of the log record (i.e. shallow copies in Go, which are
```
Comment (cijothomas) [inline specification/logs/sdk.md:341]: does the spec need to mention language specifics?

Comment (pellared) [reply specification/logs/sdk.md:341]: Requested by https://github.com/open-telemetry/opentelemetry-specification/pull/4067#pullrequestreview-2131838366 to make the specification more clear by providing examples

File: specification/logs/sdk.md [hunk 6]
```diff
@@ -317,9 +317,29 @@ therefore it SHOULD NOT block or throw exceptions.
...
+It is implementation specific whether and how the `logRecord` passed by the SDK
+is shared between the processors. Log record mutations do not have to be visible
+in next registered processors.
+
+If the `logRecord` is shared[^shared], then a `LogRecordProcessor` may freely
+modify `logRecord` only for the duration of the `OnEmit` call.
+The modifications applied on shared data are visible in next
+registered processors.
+If `logRecord` is needed after `OnEmit` returns (i.e. for asynchronous
+processing) only reads are permitted.
```
Comment (cijothomas) [inline specification/logs/sdk.md:329]:
```suggestion
processing) only reads are permitted, and it'll also reflect changes made by next processors, if any.
```

Is this correct?

File: specification/logs/sdk.md (hunk 5 above)
Comment (cijothomas) [reply specification/logs/sdk.md:341]: i am not convinced that this language is correct. The non-shared means, LogRecord must be copied. And it could be shallow copy or deep copy depending on the implementation and spec is leaving it upto implementation to decide if they should do shallow/deep copy?

File: specification/logs/sdk.md (hunk 6 above)
Comment (pellared) [reply specification/logs/sdk.md:329]: This is a portion added by @jack-berg in https://github.com/open-telemetry/opentelemetry-specification/pull/2969. I would rather not add it may lead to implementing processors which may lead to race conditions.

I could change it to something like

```suggestion
processing) only reads should be used as modifications can lead to race conditions.
```

We could also say that modifications could be done if the log record used for asynchronous processing is cloned (e.g. via an isolating processor proposed here https://github.com/open-telemetry/opentelemetry-specification/pull/4062). But I think such clarifications should be better scoped in a separate PR.

@cijothomas, @jack-berg, thoughts?

File: specification/logs/sdk.md (hunk 5 above)
Comment (pellared) [reply specification/logs/sdk.md:341]: >  And it could be shallow copy or deep copy depending on the implementation and spec is leaving it upto implementation to decide if they should do shallow/deep copy?

This is correct.

File: specification/logs/sdk.md [hunk 7]
```diff
@@ -318,9 +318,29 @@ therefore it SHOULD NOT block or throw exceptions.
...
-A `LogRecordProcessor` may freely modify `logRecord` for the duration of
-the `OnEmit` call. If `logRecord` is needed after `OnEmit` returns (i.e. for
-asynchronous processing) only reads are permitted.
+It is implementation specific whether and how the `logRecord` passed by the SDK
+is shared between the processors. Log record mutations do not have to be visible
+in next registered processors.
+
+If the `logRecord` is shared[^shared], then a `LogRecordProcessor` may freely
+modify `logRecord` only for the duration of the `OnEmit` call.
+The modifications applied on shared data are visible in next
```
Comment (MrAlias) [inline specification/logs/sdk.md:327]:
```suggestion
The modifications applied on shared data are visible in the next
```

File: specification/logs/sdk.md [hunk 8]
```diff
@@ -318,9 +318,29 @@ therefore it SHOULD NOT block or throw exceptions.
...
+registered processors.
+If `logRecord` is needed after `OnEmit` returns (i.e. for asynchronous
```
Comment (MrAlias) [inline specification/logs/sdk.md:329]:
```suggestion
If the `logRecord` is needed after `OnEmit` returns (i.e. for asynchronous
```

**Review by MrAlias — APPROVED** [review]

**Review by jack-berg — COMMENTED** [review]
I don't think this is necessary. Now that #4062 is merged, implementations like go and c++ that wish to keep log records on the stack can use the isolating processor to accomplish that, and document that users who register regular log processors should be aware of additional heap allocations. SDK implementations can build in optimizations which keep log records on the stack if only isolating processors are registered.

Additionally, I think the approach taken by C++ of stabilizing a chaining processor concept which has no basis in spec language is decidedly not correct. I understand that it was done in the name of performance, but the process we want to promote is what the go SIG is doing, where prototypes are built and used to provide feedback to evolve the spec. Stabilizing a concept then updating the spec to reflect that is the wrong order.

**Comment by pellared** [comment]
> Now that https://github.com/open-telemetry/opentelemetry-specification/pull/4062 is merged, implementations like go and c++ that wish to keep log records on the stack can use the isolating processor to accomplish that, and document that users who register regular log processors should be aware of additional heap allocations.

This is not true. If Go would provide "a regular log processors" then the additional heap allocations would be always there. This is unfortunately because of the way how Go escape analysis works and the fact that passing a pointer to an interface method results in a heap allocation. There are also other problems outlined here: https://github.com/open-telemetry/opentelemetry-go/pull/5478.

> SDK implementations can build in optimizations

The API design (pointer vs value passing) makes it impossible to build optimizations in the SDK implementation. The issue is that API design choices make influence on the performance.

**Comment by jack-berg** [comment]
> This is not true. Go would still have tendency to have more additional heap allocation. Even if we use an isolating processor - if there would be a processor that modifies e.g. a timestamp - it would lead to a heap allocation.

Why? I thought the point of the definition of isolating processor is that the log records passed to `OnEmit` has an _isolated_ log record. It seems like an SDK would provide a new isolating log processor interface, with `OnEmit` accepting an alternative log record definition which is stack allocated and which has copy-on-write mutation semantics.

> The exported API design (pointer vs value passing) makes it impossible to build optimizations in the SDK implementation. The issue is that exported API design choices make an influence on the performance.

This was a concern of mine I voiced in the spec SIG several weeks ago: trying to optimize processors while the down stream exporter design at the OTLP protocol level still strongly encourages heap allocation seems pointless. If the main point of the log bridge API / SDK is to export logs out of the application via OTLP, and the OTLP exporter can't reasonably avoid heap allocation, then who is really benefiting from this optimization work?

**Comment by pellared** [comment]
> It seems like an SDK would provide a new isolating log processor interface, with OnEmit accepting an alternative log record definition which is stack allocated and which has copy-on-write mutation semantics.

If I understand you correctly this would be true for a design looking like this: https://github.com/open-telemetry/opentelemetry-go/pull/5470. However, at this point of time, I do not like this design as I outlined [here](https://github.com/open-telemetry/opentelemetry-specification/issues/4010#issuecomment-2126709559). Do you think that this would be a better design? Notice, it still it is not compliant with the way the specification is written right now. I will do my best to also reiterate on this idea with the Go SIG.

> If the main point of the log bridge API / SDK is to export logs out of the application via OTLP, and the OTLP exporter can't reasonably avoid heap allocation, then who is really benefiting from this optimization work?

They can be use cases when people would prefer to use other exporters (allowing more throughput e.g. [user_events](https://docs.kernel.org/trace/user_events.html), [LLTng](https://lttng.org/docs/v2.13/#doc-tracing-your-own-user-application), [ETW](https://learn.microsoft.com/en-us/windows/win32/etw/about-event-tracing)). The Logs SDK design should not be limiting the performance for a single setup (even if this would be the most popular one). 

Related:
- https://github.com/open-telemetry/opentelemetry-go/pull/5578
- https://github.com/open-telemetry/opentelemetry-specification/issues/4134

**Comment by jack-berg** [comment]
> If I understand you correctly this would be true for a design looking like this: https://github.com/open-telemetry/opentelemetry-go/pull/5470. However, at this point of time, I do not like this design as I outlined https://github.com/open-telemetry/opentelemetry-specification/issues/4010#issuecomment-2126709559. Do you think that this would be a better design? Notice, it still it is not compliant with the way the specification is written right now. I will do my best to also reiterate on this idea with the Go SIG, but will also appreciate any feedback from your side.

In my opinion, the go PR#5470 is more aligned with the spirit of the spec than the proposal of this current PR. If this PR is merged, then we'll be promoting significantly diverging concepts across language implementations.

- In some languages, the modifications you make to log records are automatically visible to other processors. There's no need for chaining, and even though chaining is possible its hard to do in practice because other processors can be registered which aren't aware of the chain and operate independently.
- In other languages, modifications make a copy to the log record and you're required to call the next processor in the chain. Filtering by not calling the next entry in the chain becomes a feature and users from languages which don't do chaining are left asking how they can achieve the same affect with no good answer.

The idea behind go PR#5470 is that the way modifications are achieved is by means of returning a copy with the changes. This seems aligned with the spirit of the spec. While the spec says that `OnEmit` returns void, what if a programming language didn't support the concept of modifying a reference? The most natural thing to do would be to return the log record with the changes instead. 

Another benefit of go PR#5470 is that is that span processor could be more easily evolved alongside, retaining symmetry. This could pave a path towards a solution to #1089 / #4024, where spans can be modified on end by either modifying the reference in place with the side affects visible to other processors or by returning a copy with the modifications - based on what is idiomatic in the language. 


This type of experience isn't _too_ different from the experience in a language where you make the changes directly on the reference.

**Comment by pellared** [comment]
@jack-berg, can you share any design flaw of the proposed design apart from not being complaint with the specification and what other languages have to offer?

As you mentioned the design complaint with the specification has flaws that some you already mentioned like:
- problems with filtering
- problems with having many isolated pipelines (with interface like PR#5470 I find it hard to come up with a clean solution)
- easy to introduce a race condition
- two technical ways of composing pipelines -> passing many processors which are executed in sequence, decorating processors

I think that the spirit of the specification should be more or less like:
1. thrive for high performance
2. enable log processing of log records so that the user can setup its pipelines (modifications, filtering)
3. simple and batching processor out of the box
4. enable custom processors and exporters

I am not sure if we should follow the specification if it would result in less user-friendly and less performant design. From my experience, our users care more about the features, performance, compatibility with other OTel SDKs, tools (vua exporters, propagators), having idomatic design than being literal complaint with the specification.

Maybe an alternative could be that the @open-telemetry/technical-committee would accept the Go SDK to diverge from other languages and specification for sake of better performance and more flexible and Go idiomatic design.

**Comment by jack-berg** [comment]
> I think that the spirit of the specification should be more or less like:

I would add "symmetry with other signals where it makes sense", and "strong adherence to backwards compatibility". 

> enable log processing of log records so that the user can setup its pipelines (modifications, filtering, fan-out)

I would remove fan out and filtering as a concern of the log processors. For filtering, we should introduce a sampler concept analogous to span sampler. That way we have symmetry across signals, and dedicated tools for filtering (samplers) and mutation (processor). For fan out, while it should be possible in the SDK, it shouldn't be a high priority. We should encourage users to configure complex pipeline in the collector instead of implementing the tooling for 11 different languages. We have spec language suggesting this [here](https://github.com/open-telemetry/opentelemetry-specification/blob/3efed53cc1842b2a1d74e69c468e268649cf1d34/specification/logs/sdk.md#L382-L384).

> I am not sure if we should follow the specification if it would result in less user-friendly and less performant design. 

The go PR#5470 proposal shouldn't be any less performant and the the less user-friendly is subjective. I'd argue that symmetry / familiarity across languages is a key element of being user friendly.

The idea of formalizing chaining processors is a significant change to design of SDK processors in the spec, and should be considered against the span processor and sampler design. I actually prefer the chaining design to the current processor design, but am opposed to fragmenting the design with incremental changes.

> Maybe an alternative could be that the @open-telemetry/technical-committee would accept the Go SDK to diverge from other languages and specification for sake of better performance, more flexible and Go idiomatic design.

My opinion is just one of many who have the ability to weigh in and influence this conversation. I'm not sure where others stand on this topic, but its certainly possible that others align with you're point of view. While I don't prefer the proposal in this PR, I wouldn't block it.

**Comment by pellared** [comment]
> The go PR#5470 proposal shouldn't be any less performant

It is still (a little) less performant because there is more copying on the stack. In given benchmarks it shows on average ~9% CPU overhead on my machine.

File: specification/logs/sdk.md (hunk 6 above)
Comment (pellared) [reply specification/logs/sdk.md:330]: I will create a separate PR for this.

Comment (pellared) [reply specification/logs/sdk.md:330]: https://github.com/open-telemetry/opentelemetry-specification/issues/4141

**Comment by cijothomas** [comment]
I had some confusion as to why OTel Go does not allow changes to be visible to the next processor, even though OTel Go want to achieve high performance (same goal as other languages like OTel Rust, but those languages allow changes to be visible to the next processor.) It has to do with the way OTel Go allocates LogRecord. Pasting below my understanding for easy reference for those who are curious (Thanks @pellared who explained OTel Go behavior in detail).

OTel .NET
```
// LogRecord is always on the heap.
var ptrLogRecord = new LogRecord {}; 

// pass pointer to the LogRecord to processors
// any changes made to LogRecord are visible to next processors
// The above means that OTel .NET will always have a heap allocation for logs,
// but the SDK does pooling to mitigate.
processor.OnEmit(ptrLogRecord); 
```
																
OTel Rust				
```
// LogRecord is on the *stack*
var ptrLogRecord = LogRecord {}; 

// pass pointer to the LogRecord to processors. (This is the key diff. with OTel Go, as Go would force heap alloc, if pointer is passed!)
// any changes made to LogRecord are visible to next processors.
// Processor *cannot* store the ptrLogRecord beyond the OnEmit() call. (Rust compiler enforces this :) )
// If the processor must store the LogRecord beyond that (eg: BatchProcessor),
// it must clone and store the cloned version.
// When cloned like this, any changes done by next processors won't be visible.
// The above allows OTel Rust to have 0 heap alloc when used with sync exporters like etw.
```

OTel Go
```
// LogRecord is on the *stack*, as long it is passed by Value to processor.
var logRecord = LogRecord {};

// pass LogRecord copy, (yes, there is copy cost involved) *NOT* the pointer!
// because copy is given to Processor,
// mutations are not visible to next Processor.
// Changing this to accept a ptrLogRecord will
// force Go's escape analysis to trigger
// allocating LogRecord on the heap instead.
// But that'll make changes visible to next processors.
processor.OnEmit(logRecord); 
```

Given this, I think it is okay to call out in spec that LogRecord changes may/may not be visible, depending on the implementation. Alternatively, I am also fine to allow OTel Go to deviate from existing spec wording, and document it in OTel-Go repo. And OTel Go repo can provide solutions to end-users who want to leverage `LogProcessor` to enrich/redact/modify LogRecord, by leveraging the IsolatingLogProcessor/Decorator.

Also, though OTel Go copies LogRecord, it is shallow copy only, so if user has more than 5 attributes, those extra attributes are not copied, so changes to them will be visible across processors!

**Comment by pellared** [comment]
> Alternatively, I am also fine to allow OTel Go to deviate from existing spec wording, and document it in OTel-Go repo. 

We strongly prefer to make it possible via specification. We

**Comment by pellared** [comment]
> Maybe I could reword so that "shared" is recommended and "not-shared" allowed in specific languages as an exception e.g. for performance reasons.

PR updated

> The Go SIG is committed to not block future changes of the specification even if some would make it harder to implement.

Does it need to be formalized?

**Review by jmacd — APPROVED** [review]

**Comment by pellared** [comment]
**SIG meeting notes:** needs more reviews.

**Comment by pellared** [comment]
> Filtering by not calling the next entry in the chain becomes a feature and users from languages which don't do chaining are left asking how they can achieve the same affect with no good answer.

The answer is in SDK documentation, examples etc. Most users are not using the specification as a reference, but the SDK and opentelemetry.io documentation.

I want also to call out that even know all existing SDK implementations can do filtering by wrapping/chaining/decorating in all languages. People are doing it as it is technically possible. Nothing in the SDKs' docs or specification discourages from doing it.

**Comment by jack-berg** [comment]
> I want also to call out that even know all existing SDKs users can do filtering by wrapping/chaining/decorating. People are doing it as it is technically possible. Nothing in the SDKs' docs or specification discourages from doing it.

Yes but we design tools on top of the specific patterns we promote in the SDK. For example, the file configuration model schema reflects the SDK log processor design. A `.logger_provider_.processors` is a sequence of processors with the expectation that the SDK calls all of them. A chaining design would mean that there is no guarantee that all are called and the cross-language support of file configuration is diminished. 

> Having a inefficient and less flexible design will discourage many Go developers from using SDK Processors and they would use logging libraries processing capabilities instead.

PR#5470 appears to be marginally less efficient while retaining consistency with the spec design. I suggest go adopts this proposal, and separately pursues evolving the spec with a chaining processor design, holistically considering how it would work with spans, logs, and maybe metrics. 

> If we have a less flexible design (yet performance wise acceptable) design like in PR#5470 then I am certain that we would get negative feedback from the Go community that the design is bad.

I think we need to separate the issues. PR#5470 represents a reasonable way to design the go log SDK without jeopardizing performance. Separately, there's a desire for SDKs to have a intuitive extension point for data pipelining tasks like filtering, transforming, fan-out, etc. Pursuing PR#5470 allows you to land a stable go log SDK with less capable data pipelining tools, but which can evolve and improve in the future if someone can successfully drive a new chaining processor design. 

Some users will be disappointed, preferring what they view as a more elegant chaining processor design from the start, but their argument can be effectively countered: Its possible for them to do chaining from the start with the PR#5470, so why do they care that the ergonomics are slightly less than optimal? OpenTelemetry is first and foremost a standards effort, and the ergonomics they want are in conflict with what is written and implemented. Better ergonomics means changing the spec, and changing the spec means taking a holistic approach to design. Additionally, the vast majority of users will not want to use the Otel log SDK to do anything more than export via OTLP, and we explicitly encourage doing processing the collector instead of SDKs to limit the amount of duplicated effort across the 11 language implementations. The set of users who stand to be impacted by a log SDK design with is performance with less than optimal ergonomics has to be quite small.

**Comment by trask** [comment]
as someone who'd like to see chaining processors become a top-level feature across both spans and logs, I'd personally prefer not to carve out an exception here only for Go, and instead solve it for all languages

**Comment by pellared** [comment]
>  A chaining design would mean that there is no guarantee that all are called and the cross-language support of file configuration is diminished.

Can you an example of a configuration that we won't be able to support using chaining design? Can you give my any example of any setup that is possible with "shared" log record processing that is not possible to setup with chaining processors?

> explicitly encourage doing processing the collector instead of SDKs to limit the amount of duplicated effort across the 11 language implementations

I do not say that SDKs need to provide more processor implementations. From what you say it means that the whole idea of supporting custom processing is not important.

> The set of users who stand to be impacted by a log SDK design with is performance with less than optimal ergonomics has to be quite small.

Does it mean that we should ignore these users? 

I have not come up with a single example of a processing pipeline that can setup in other languages that cannot be setup using chained processors.

One of the ideas behind adding an isolating processor is that other languages can use it to make a chaining processor setup.

**Comment by pellared** [comment]
> as someone who'd like to see chaining processors become a top-level feature across both spans and logs, I'd personally prefer not to carve out an exception here only for Go, and instead solve it for all languages

With isolating processor I think all languages are able setup chaining processor pipelines even now. Pseudocode:

```
provider.addProcessor( isolatingProcessor( redactingProcessor( batchProcessor ( otlp exporter ) ) ) )

provider.addProcessor( isolatingProcessor( ( simpleProcessor ( stdout exporter ) ) ) 
```

**Comment by jack-berg** [comment]
> At some point, I was thinking about providing a new addChainingProcessor functionality to the log provider. However, I feel that it would be an overkill and unnecessarily increase the complexity of the SDKs (and later configuration etc). I prefer to evolve to the existing processing design.

This is what I'm advocating for, but adding to LoggerProvider, TracerProvider, and maybe MeterProvider. The idea of a chaining processor is different enough (and valuable enough) that we should pursue it for all the signals, not just logging. Leave the current notion of processors and all their flaws alone, and evolve a new concept.

**Comment by pellared** [comment]
> This is what I'm advocating for

Why not specifying that users can chain existing processors and use the isolating processor? The same can be done for tracing.

**Comment by dashpole** [comment]
I agree with @jack-berg's concerns here, and don't think we should introduce an exception that allows languages to functionally change how processors interact. I don't think https://github.com/open-telemetry/opentelemetry-go/pull/5470 complies with the current spec either. A heap allocation on the hot path is probably not going to be acceptable for a lot of Go users, though, and will hinder the adoption of the Go SDK until we address it.  But it does seem like the best path is to introduce a new concept, rather than carve an exception out of the existing concept.

It seems like we all agree that a chaining processor would be worth introducing across signals. Is there a process for deprecating things in the spec?  My only concern with a new processor with a slightly different design is that it would confuse users.

File: specification/logs/sdk.md [hunk 9]
```diff
@@ -318,9 +318,24 @@ therefore it SHOULD NOT block or throw exceptions.
...
+conditions.
+
+In exceptional cases, the SDK may not share the `logRecord` between processors.
+For example, for better performance, the Go SDK passes shallow copies of the
+`logRecord` to the processors, and these copies are stack allocated.
+If it were passed by pointer, the Go compiler would allocate them on
+the heap and increase the GC pressure.
+If a modification is needed in another processor,
+then the processor making the modification needs to call the other processor's
+`OnEmit` and pass the locally modified `logRecord`.
```
Comment (jmacd) [inline specification/logs/sdk.md:338]: Given the concerns we've heard about this in the past -- about confusion over the two different models being optionally available in each SDK -- what if we used different method names to capture the difference? For example, OnEmit() could call the shared-instance pipeline with traditional semantics, and OnEmitCopy() could call the copied-instance pipeline with new semantics? (I couldn't think of a better name!)

**Comment by pellared** [comment]
The Go SIG decided that the Go Logs SDK does not have to be an exception. I change the PR so that it only clarifies the specification.

File: specification/logs/sdk.md [hunk 10]
```diff
@@ -324,6 +324,11 @@ therefore it SHOULD NOT block or throw exceptions.
 
 **Returns:** `Void`
 
+The passed `logRecord` MUST be shared between registered [`LogRecordProcessors`](#logrecordprocessor).
+This means that each processor receives the same instance of the `logRecord`
+(i.e. shared by a pointer).
+The applied modifications are visible in the next registered processors.
```
Comment (cijothomas) [inline specification/logs/sdk.md:330]: wording suggestion: Can we make the spec only mandate that the changes are visible to the next processor? whether its achieved via pointer or something else is an implementation detail, best omitted from spec language.

Comment (pellared) [reply specification/logs/sdk.md:330]: Sure 👍

Comment (pellared) [reply specification/logs/sdk.md]: Makes sense. PTAL https://github.com/open-telemetry/opentelemetry-specification/pull/4067/commits/711ca57ac7a3bbbdd271de20b20d8d250db92a3b

**Review by cijothomas — APPROVED** [review]

**Review by jsuereth — APPROVED** [review]
Given all the discussions and options here, I think this is the right path forward with what we have today.

**Review by dashpole — APPROVED** [review]

**Comment by cijothomas** [comment]
@pellared Can you resolve the conflict? This looks like has enough approvals, so should be good to merge.

**Comment by tigrannajaryan** [comment]
> @pellared Can you resolve the conflict? This looks like has enough approvals, so should be good to merge.

I think @pellared is out. We may need someone else to take over the PR.

**Comment by cijothomas** [comment]
> > @pellared Can you resolve the conflict? This looks like has enough approvals, so should be good to merge.
> 
> I think @pellared is out. We may need someone else to take over the PR.

Looks like he just appeared to fix conflict :) 

@jmacd @arminru Tagging to request review as PR seems assigned to you both!

**Comment by cijothomas** [comment]
@open-telemetry/technical-committee Could you suggest next steps for moving this forward? There are approvals already, but no further movement and Bot will mark this stale again soon. 
It is totally fine if this is just waiting for more eyes to review and/or people are busy.

**Comment by carlosalberto** [comment]
Let's mention this in our Spec call today and otherwise we should merge this in the next 1-2 days, as we have left it open for a little while and no concerns have been raised.

**Comment by cijothomas** [comment]
> Let's mention this in our Spec call today and otherwise we should merge this in the next 1-2 days, as we have left it open for a little while and no concerns have been raised.

Thanks!
I won't be able to join the spec call, most likely:(
 Could you help mention this in the call today?

**Comment by carlosalberto** [comment]
Just to be super clear: what's the final opinion from the @open-telemetry/cpp-maintainers on this topic?

File: specification/logs/sdk.md [hunk 11]
```diff
@@ -324,6 +324,9 @@ therefore it SHOULD NOT block or throw exceptions.
 
 **Returns:** `Void`
 
+For a `LogRecordProcessor` registered directly on SDK `LoggerProvider`,
+the `logRecord` mutations MUST be visible in next registered processors.
+
```
Comment (lalitb) [inline specification/logs/sdk.md:329]: As noted in one of the [comments](https://github.com/open-telemetry/opentelemetry-specification/issues/4010#issuecomment-2096577303), otel-cpp does not support this feature. Due to performance considerations, the inherent design of logs and traces in the SDK avoids creating intermediate SpanData or LogRecord objects. This means that mutation is not possible either.

**Review by trask — APPROVED** [review]
I think LogRecordProcessors which stamp additional log attributes prior to export is a common use case (e.g. I have implemented two of them [here](https://github.com/microsoft/ApplicationInsights-Java/blob/main/agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/AzureFunctionsLogProcessor.java) and [here](https://github.com/microsoft/ApplicationInsights-Java/blob/main/agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/AzureMonitorLogProcessor.java))

**Comment by pellared** [comment]
@open-telemetry/technical-committee, can you please check whether this PR is good to be merged?
